internal/cloudsql/instance.go

```diff
@@ -0,0 +1,209 @@
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudsql
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"regexp"
+	"sync"
+	"time"
+
+	"golang.org/x/net/proxy"
+	sqladmin "google.golang.org/api/sqladmin/v1beta4"
+)
+
+var (
+	// Instance connection name is the format <PROJECT>:<REGION>:<INSTANCE>
+	// Additionally, we have to support legacy "domain-scoped" projects (e.g. "google.com:PROJECT")
+	connNameRegex = regexp.MustCompile("([^:]+(:[^:]+)?):([^:]+):([^:]+)")
+
+	// defaultKeepAlive is the default keep alive value on connections created in this package.
+	defaultKeepAlive = 30 * time.Second
+)
+
+// connName represents the "instance connection name", in the format "project:region:name". Use the
+// "parseConnName" method to initialize this struct.
+type connName struct {
+	project string
+	region  string
+	name    string
+}
+
+func (c *connName) String() string {
+	return fmt.Sprintf("%s:%s:%s", c.project, c.region, c.name)
+}
+
+// parseConnName initializes a new connName struct.
+func parseConnName(cn string) (connName, error) {
+	b := []byte(cn)
+	m := connNameRegex.FindSubmatch(b)
+	if m == nil {
+		return connName{}, fmt.Errorf("invalid instance connection name - expected PROJECT:REGION:ID")
+	}
+
+	c := connName{
+		project: string(m[1]),
+		region:  string(m[3]),
+		name:    string(m[4]),
+	}
+	return c, nil
+}
+
+// refreshResult is a pending result of a refresh operation of data used to connect securely. It should
+// only be initialized by the Instance struct as part of a refresh cycle.
+type refreshResult struct {
+	md     metadata
+	tlsCfg *tls.Config
+	err    error
+
+	// timer that triggers refresh, can be used to cancel.
+	timer *time.Timer
+	// indicates the struct is ready to read from
+	ready chan struct{}
+}
+
+// Cancel prevents the instanceInfo from starting, if it hasn't already started. Returns true if timer
+// was stopped successfully, or false if it has already started.
+func (i *refreshResult) Cancel() bool {
+	return i.timer.Stop()
+}
+
+// Wait blocks until the refreshResult attempt is completed.
+func (i *refreshResult) Wait(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-i.ready:
+		return i.err
+	}
+}
+
+// Instance manages the information used to connect to the Cloud SQL instance by periodically calling
+// the Cloud SQL Admin API. It automatically refreshes the required information approximately 5 minutes
+// before the previous certificate expires (every 55 minutes).
+type Instance struct {
+	connName
+	client *sqladmin.Service
+	key    *rsa.PrivateKey
+
+	resultGuard sync.RWMutex
+	cur         *refreshResult
```

Comment:
Would it be more clear to have a brief docstring above cur and next to describe what they are?